modules/compiler/utils/source/degenerate_sub_group_pass.cpp

```diff
@@ -50,59 +50,53 @@ namespace {
+    return GroupOp->isSubGroupScope();
   }
 
-  return Fcn->getName() == compiler::utils::MuxBuiltins::sub_group_barrier;
+  return Builtin.ID == compiler::utils::eMuxBuiltinSubGroupBarrier;
 }
 
 /// @brief Helper for building the symbol name of the mangled work-group builtin
 /// corresponding to the sub-group builtin.
 ///
-/// @param[in] SubgroupBuiltin sub-group builtin to map to a work-group builtin.
-/// @param[in] Ctx Context used to create llvm objects against.
-/// @param[in] M in which these builtins are created.
-///
 /// @return The mangled work-group builtin corresponding to `SubgroupBuiltin`.
-std::string lookupWGBuiltin(StringRef SubgroupBuiltin, LLVMContext *Ctx,
-                            Module *M) {
+Function *lookupWGBuiltin(Function &SubgroupF, Module &M) {
   // We must handle the case where we're replacing a __mux_sub_group_barrier
   // with a __mux_work_group_barrier. Our 'demangleName' API works differently
   // with non-mangled builtin names and returns an empty string. Just work
   // around it specifically.
-  if (SubgroupBuiltin == compiler::utils::MuxBuiltins::sub_group_barrier) {
-    return compiler::utils::MuxBuiltins::work_group_barrier;
+  auto SubgroupFName = SubgroupF.getName();
+  auto SubgroupFTy = SubgroupF.getFunctionType();
+  if (SubgroupFName == compiler::utils::MuxBuiltins::sub_group_barrier) {
+    auto *WorkgroupF =
+        M.getOrInsertFunction(compiler::utils::MuxBuiltins::work_group_barrier,
+                              SubgroupFTy)
+            .getCallee();
+    return cast<Function>(WorkgroupF);
   }
 
-  compiler::utils::NameMangler Mangler(Ctx);
-  SmallVector<Type *, 4> ArgumentTypes;
-  SmallVector<compiler::utils::TypeQualifiers, 4> Qualifiers;
-
-  const auto DemangledName = std::string(
-      Mangler.demangleName(SubgroupBuiltin, ArgumentTypes, Qualifiers));
-
-  const auto WorkGroupBuiltinName = "work" + DemangledName.substr(3);
-  // We have to special case broadcast here since the sub-group version takes
-  // a single uint but we need to map this to the 3D work-group version which
-  // takes a size_t.
-  if (std::string::npos != WorkGroupBuiltinName.find("broadcast")) {
-    // Here we are mapping Tj -> Tmmm for any type T (assuming size_t is
-    // unsigned long). So first remap the existing type.
-    auto *const sizeTy = compiler::utils::getSizeType(*M);
-    ArgumentTypes.back() = sizeTy;
-
-    // Then we need to push back two more size_ts for the Y and Z arguments.
-    for (unsigned i = 0; i < 2; ++i) {
-      ArgumentTypes.push_back(sizeTy);
-      Qualifiers.push_back(compiler::utils::eTypeQualNone);
+  std::string WorkgroupFName = "__mux_work" + SubgroupFName.substr(9).str();
+
+  // Work-group builtins have an extra 'barrier id' as the first parameter.
+  SmallVector<Type *, 4> ArgTys;
+  ArgTys.push_back(IntegerType::getInt32Ty(M.getContext()));
+  ArgTys.push_back(SubgroupF.getArg(0)->getType());
+  // Broadcasts have different IDs
```

Comment:
as an additional ID?

Reply:
I've clarified this comment, thanks. It was a bit confusing with respect to the 'barrier IDs' mentioned above.